server/blob/blob-deduplicating/src/test/scala/org/apache/james/server/blob/deduplication/GCPropertiesTest.scala

```diff
@@ -0,0 +1,215 @@
+/***************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+package org.apache.james.server.blob.deduplication
+
+import java.nio.charset.StandardCharsets
+
+import com.google.common.hash
+import org.apache.james.blob.api.BlobId
+import org.apache.james.server.blob.deduplication.Generators.{OnePassGCTestParameters, TestParameters}
+import org.scalacheck.Prop.forAll
+import org.scalacheck.Test.Parameters
+import org.scalacheck.{Arbitrary, Gen, Properties, Shrink}
+
+case class GenerationAwareBlobId(generation: Generation, hash: String) extends BlobId {
+  override def asString(): String = s"${generation}_$hash"
+}
+
+object Generators {
+
+  // generate a sequence of Generations with monotonic numeric ids
+  // 80% of the time, the generation id is not incremented
+  // 19% of the time, the generation id is incremented by 1
+  // 1% of the time, the generation id is incremented by 2
+  // i.e. (0, 0, 0, 2, 3, 3, 4, 5, 5, 5, 5)
+  def nextGenerationsGen(previousGeneration: Generation): Gen[Generation] =
+    Gen.frequency((80, Gen.const(0l)), (19, Gen.const(1l)), (1, Gen.const(2l))).map(previousGeneration.next)
+
+  val externalIDGen: Gen[ExternalID] = Gen.uuid.map(uuid => ExternalID(uuid.toString))
+
+  def referenceGen(generation: Generation, hash: String): Gen[Reference] = for {
+    externalId <- externalIDGen
+  } yield Reference(externalId, GenerationAwareBlobId(generation, hash), generation)
+
+  case class ReferenceAccumulator(dereferenced: Set[Reference], existing: Set[Reference]) {
+    def addDeletion(dereference: Reference) = this.copy(dereferenced = dereferenced ++ Set(dereference))
```

Comment:
It is allowed to dereference something not referenced?

Reply:
No, and generators ensure that assertion